doc/design/proposal-synchronous-hooks.md

```diff
@@ -0,0 +1,258 @@
+4. It seems useful to allow the results returned by the hooks to be partial i.e. hooks don't have to clone the result returned by the next (default) hook to override it, instead they only need to return an object with properties that they wish to override. This can save the overhead of excessive clones.
+5. It seems `shortCircuit` is not really necessary if hooks can just choose to not call the next hook?
+
+## Hooks
+
+### `resolve`: from specifier to url
+
+```js
+/**
+ * @typedef {{
+ *   parentURL?: string,
+ *   conditions?: string[],
+ *   importAttributes?: object[]
+ * }} ModuleResolveContext
+ */
+/**
+ * @typedef {{
+ *   url?: string,
+ *   format?: string
+ * }} ModuleResolveResult
+ */
+/**
+ * @param {string} specifier
+ * @param {ModuleResolveContext} context
+ * @param {(specifier: string, context: ModuleResolveContext) => ModuleResolveResult} nextResolve
+ * @returns {ModuleResolveResult}
+ */
+function resolve(specifier, context, nextResolve) {
+  if (shouldOverride(specifier, context.parentURL)) {
+    const url = customResolve(specifier, context.parentURL);
+    return { url };
+  }
+
+  const resolved = nextResolve(specifier, context);
+  if (resolved.url.endsWith('.zip')) {
+    return { format: 'zip' };
+  }
+  return {};  // no override
+}
+```
+
+Notes:
+
+1. Example use case: yarn pnp ([esm hooks](https://github.com/yarnpkg/berry/blob/master/packages/yarnpkg-pnp/sources/esm-loader/hooks/resolve.ts), [cjs hooks](https://github.com/yarnpkg/berry/blob/master/packages/yarnpkg-pnp/sources/loader/applyPatch.ts))
+2. `importAttributes` are only available when the module request is initiated with `import`.
+3. For the CJS loader, `Module._cache` is keyed using file names, and it's used in the wild (usually in the form of `require.cache`) so changing it to key on URL would be breaking. We'll need to figure out another way to map it with an url.
+   1. It might be non-breaking to maintain an additional map on the side for mapping the same url with different searches and hashes to a different module instance, or leave `Module._cache` as an alias map for the first module instance mapped by the URL minus search/hash. If in the same realm, `Module._cache` gets directly accessed from the outside, we can emit a warning about the incompatibility. Down the road this may require making `Module._cache` a proxy to propagate delete requests to the actual URL-based map.
+   2. Changing CJS modules to be mapped with URL will have performance consequences from but might also help with the hot module replacement use case. Note that for parent packages with `exports` or `imports` in their `package.json` the URL conversion is already done (and not even cached) even in the CJS loader.
+
+### `load`: from url to source code
+
+```js
+/**
+ * @typedef {{
+ *   format?: string,
+ *   conditions?: string[],
+ *   importAttributes?: object[]
+ * }} ModuleLoadContext
+ */
+/**
+ * @typedef {{
+*   format?: string,
+*   source: string | Buffer
```

Comment:
Could we add an optional field `sourcemap` to enable customize sourcemap loading as well? 

Ref: https://github.com/nodejs/node/issues/51781

Reply:
I think `load` (url -> source code) may not be the best timing for this because we are also looking into letting V8 to parse the magic comments to get the URLs. And load happens before parsing.

Reply:
Actually for CJS, if we span the exports hook around evaluation, I think that would be a good timing for this override. Also it can potentially address the problem of contextual access of `exports` not being mutable if the hook return a completely new `exports`, since they can invoke `nextExports` with that new exports object if the default exports hook runs the CJS wrapper (and passing in the contextual `exports`).

Reply:
Added a section at the bottle to discuss sourcemap (but it won't be in the initial implementation).